homeassistant/components/shell_command/__init__.py

```diff
@@ -132,7 +124,18 @@ async def async_service_handler(service: ServiceCall) -> ServiceResponse:
                 "Error running command: `%s`, return code: %s", cmd, process.returncode
             )
 
-        return service_response
+        if service.return_response:
+            service_response: JsonObjectType = {
+                "stdout": "",
+                "stderr": "",
+                "returncode": process.returncode,
+            }
+            if stdout_data:
+                service_response["stdout"] = stdout_data.decode("utf-8").strip()
+            if stderr_data:
+                service_response["stderr"] = stderr_data.decode("utf-8").strip()
```

Comment:
Will the response in all cases be utf-8? Will there be use cases where people can get other types of data to do other stuff with?

If we always expect it to be utf-8, can we throw something to alert the user that they did something wrong?

Reply:
1. Not guaranteed, but it is the implicit default anywhere where subprocesses are used.
2. Any output imaginable could pass through here, but as service response types are restricted to text we cannot handle all types. To avoid throwing errors on non-text when no output is requested, the output handling has been made conditional. If output is requested, text is required.

I'll add some additional error logging in case this throws to provide additional context for the user.

Reply:
Euh, doesn't `_LOGGER.exception` take an extra argument to substitute `%s`?

And do you if errors thrown by a service call gets linked to the place where the service call gets done? So for example you have 5 automations doing some kind of shell command and one of them fails in the middle of the night, can you see which one of those 5 it could be? If not, then I think the executed command might be useful as well.

Reply:
Damn, forgot to add the command -_-, 

Not sure if I understand your question correctly, but in the PR of the originally added feature it was discussed that whenever a returning service causes an error it should `raise` an exception instead of silently returning nothing. This is to ensure tracability. How this is handled internally inside HA I am not sure. But based on some testing by calling an erroring `shell_command` inside an automation, my logging does show that automation as having caused an error.

Reply:
Ah yes that answers my question. Thanks!